File: app/src/main/java/com/hhy/game/snowman/GameThread.java
package com.hhy.game.snowman;

import com.hhy.game.snowman.game.Game;
import com.hhy.game.snowman.game.Obstacle;

public class GameThread extends Thread {
    private Game game;
    private volatile boolean running = true;
    private final int FRAME_RATE = 1000;

    GameThread(Game game) {
        this.game = game;
    }

    @Override
    public void run() {
        long lastTime = System.currentTimeMillis();

        // Game loop
        while (running) {
            long now = System.currentTimeMillis();
            long elapsed = now - lastTime;

            if (elapsed < FRAME_RATE) {
                game.update(elapsed);
                game.draw();
            }

            Obstacle lastObs = game.getLastObs();
            if (lastObs.getCoordinates().getX() + lastObs.getWidth() <= game.getScreen().width() / 16 * 13) {
                game.generateObstacle();
            }

            lastTime = now;
        }
    }

    void shutdown() {
        running = false;
    }
}

File: app/src/main/java/com/hhy/game/snowman/GameThread.kt
package com.hhy.game.snowman

import com.hhy.game.snowman.game.Game

class GameThread internal constructor(private val game: Game) : Thread() {

    @Volatile
    private var running = true
    private val FRAME_RATE = 1000
    override fun run() {
        var lastTime = System.currentTimeMillis().toInt()

        // Game loop
        while (running) {
            val now = System.currentTimeMillis().toInt()
            val elapsed = now - lastTime
            if (elapsed < FRAME_RATE) {
                game.update(elapsed)
                game.draw()
            }
            val lastObs = game.getLastObs()
            if (lastObs.coordinates.x + lastObs.width <= game.getObstacleDistance()) {
                game.generateObstacle()
            }
            lastTime = now
        }
    }

    fun shutdown() {
        running = false
    }

}